Expose some Parquet per-column configuration options via the python API

etseidl opened the pull request:
## Description
Several recent PRs (#15081, #15411, #15600) added the ability to control some aspects of Parquet file writing on a per-column basis. During discussion of #15081 it was [suggested](https://github.com/rapidsai/cudf/pull/15081#issuecomment-1979731930) that these options be exposed by cuDF-python in a manner similar to pyarrow. This PR adds the ability to control per-column encoding, compression, binary output, and fixed-length data width, using fully qualified Parquet column names. For example, given a cuDF table with an integer column 'a', and a `list<int32>` column 'b', the fully qualified column names would be 'a' and 'b.list.element'.

Addresses "Add cuDF-python API support for specifying encodings" task in #13501.

<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

etseidl commented:
@GregoryKimball does this do what you wanted?

vuule commented:
/ok to test

vuule commented:
/ok to test

mhaseeb123 reviewed (commented):
Partial review: The changes look good so far. Thanks for the effort, Ed. I think we would benefit from some tests using `write_parquet` function and the `ParquetWriter` to write data and then using pyarrow parquet reader to read and verify correct encodings for each column. Thank you!

etseidl commented:
> I think we would benefit from some tests using `write_parquet` function and the `ParquetWriter` to write data and then using pyarrow parquet reader to read and verify correct encodings for each column.

@mhaseeb123 I added some parameterization to at least verify that all the valid Parquet encodings work on the write side. There is already testing elsewhere to verify encoding interoperability with pyarrow/pandas.[1] I also added a check that the column that should still be compressed actually is.

The option set for `ParquetWriter` has fallen far behind the `DataFrame.to_parquet()` path. Bringing the former up to date is beyond scope for this PR IMHO.

[1] With the exception of BYTE_STREAM_SPLIT, which was only recently fully implemented in arrow 16. This too can be addressed in a follow up PR.

mhaseeb123 commented on python/cudf/cudf/tests/test_parquet.py:
```diff
@@ -2870,6 +2870,82 @@ def flba(i):
+    # DICTIONARY and USE_DEFAULT should both result in a PLAIN_DICTIONARY encoding in parquet
+    encoding_name = (
+        "PLAIN_DICTIONARY"
+        if encoding == "DICTIONARY" or encoding == "USE_DEFAULT"
```
might be a naive question but do we have a check to handle the case where a v2 encoding is requested with `use_header="1.0"`

etseidl replied:
We do now :wink:. Up until now, the v2 encodings have only been accessible via defaults based on the header version. With this change any encoding valid for the datatype can be chosen, and that's perfectly ok spec wise.

mhaseeb123 reviewed (commented):
Partial review. Just a minor comment. Looks good otherwise

mhaseeb123 reviewed (approved):
Looks good to me!

mhaseeb123 commented:
/ok to test

vyasr commented on python/cudf/cudf/_lib/pylibcudf/libcudf/io/types.pxd:
```diff
@@ -57,6 +57,18 @@ cdef extern from "cudf/io/types.hpp" \
         ADAPTIVE = 1,
         ALWAYS = 2,
 
+    ctypedef enum column_encoding:
```
Rather than a C-style enum, our newer Cython code uses C++ scoped enums, which Cython now supports. [Here's an example](https://github.com/rapidsai/cudf/blob/branch-24.06/python/cudf/cudf/_lib/pylibcudf/libcudf/binaryop.pxd#L14).

etseidl replied:
Thanks for the pointer!

vyasr reviewed (changes requested):
Overall looks great, thanks! One change to make then this should be good to go.

vuule commented:
/ok to test

GregoryKimball commented:
@galipremsagar Would you please check how `cudf.pandas` runs when `column_encoding` is specified?

The user command would be something like `df.to_parquet(columns_encoding='DELTA_BYTE_ARRAY')`.

By the way, what would `cudf.pandas` do if the user specified an `engine` such as:
`df.to_parquet(engine='pyarrow', columns_encoding='DELTA_BYTE_ARRAY')`

galipremsagar commented:
> @galipremsagar Would you please check how `cudf.pandas` runs when `column_encoding` is specified?
> 
> The user command would be something like `df.to_parquet(columns_encoding='DELTA_BYTE_ARRAY')`.

This will invoke cudf parquet writer.

> By the way, what would `cudf.pandas` do if the user specified an `engine` such as: `df.to_parquet(engine='pyarrow', columns_encoding='DELTA_BYTE_ARRAY')`?

This will invoke pyarrow's parquet writer through cudf code, but ignore `columns_encoding`

galipremsagar commented:
/okay to test

vyasr reviewed (approved):
LGTM

vuule commented:
/merge
